File: ios/CircleBoxSDK/Sources/CircleBoxSDK/CircleBoxRuntime.swift
import Foundation

final class CircleBoxRuntime {
    private let stateQueue = DispatchQueue(label: "com.circlebox.sdk.runtime")
    private let fileStore: CircleBoxFileStore
    private let environmentProvider: () -> CircleBoxEnvironment

    private var environment: CircleBoxEnvironment
    private var config: CircleBoxConfig = .default
    private var ringBuffer: CircleBoxRingBuffer<CircleBoxEvent> = CircleBoxRingBuffer(capacity: 50)
    private var monitors: CircleBoxSystemMonitors?
    private var crashHandler: CircleBoxCrashHandler?
    private var signalCrashHandler: CircleBoxSignalCrashHandler?
    private var started = false

    init(
        fileStore: CircleBoxFileStore = CircleBoxFileStore(),
        environmentProvider: @escaping () -> CircleBoxEnvironment = CircleBoxEnvironment.current
    ) {
        self.fileStore = fileStore
        self.environmentProvider = environmentProvider
        self.environment = environmentProvider()
    }

    func start(config: CircleBoxConfig) {
        // Startup is idempotent: only the first call wins.
        let shouldStart: Bool = stateQueue.sync {
            guard !started else { return false }
            started = true
            self.config = config
            environment = environmentProvider()
            ringBuffer = CircleBoxRingBuffer(capacity: config.bufferCapacity)
            return true
        }

        guard shouldStart else { return }

        recoverPendingFromSignalMarkerIfNeeded()

        let monitors = CircleBoxSystemMonitors(config: config, fileStore: fileStore) { [weak self] type, severity, attrs, thread in
            self?.record(type: type, severity: severity, attrs: attrs, thread: thread)
        }
        monitors.start()
        self.monitors = monitors

        let crashHandler = CircleBoxCrashHandler { [weak self] details in
            self?.handleCrash(details: details)
        }
        crashHandler.install()
        self.crashHandler = crashHandler

        if config.enableSignalCrashCapture,
           let markerPath = fileStore.signalMarkerPathCString(),
           let signalCrashHandler = CircleBoxSignalCrashHandler(markerPath: markerPath) {
            signalCrashHandler.install()
            self.signalCrashHandler = signalCrashHandler
        }

        record(type: "sdk_start", severity: .info, attrs: ["buffer_capacity": String(config.bufferCapacity)], thread: .main)
        writeCheckpointBestEffort()
    }

    func breadcrumb(message: String, attrs: [String: String]) {
        var final = attrs
        final["message"] = message
        record(type: "breadcrumb", severity: .info, attrs: final, thread: .main)
    }

    func exportLogs(formats: Set<CircleBoxExportFormat>) throws -> [URL] {
        // Pending crash reports (written in a previous process) take precedence.
        let pendingEnvelope = fileStore.readPendingEnvelope()
        let contextDefaults: (CircleBoxExportSource, CircleBoxCaptureReason) = pendingEnvelope == nil
            ? (.liveSnapshot, .manualExport)
            : (.pendingCrash, .startupPendingDetection)
        let envelope = normalizedEnvelope(
            pendingEnvelope ?? snapshotEnvelope(exportSource: .liveSnapshot, captureReason: .manualExport),
            fallbackExportSource: contextDefaults.0,
            fallbackCaptureReason: contextDefaults.1
        )

        var urls: [URL] = []
        let orderedFormats = CircleBoxExportFormat.allCases.filter(formats.contains)

        for format in orderedFormats {
            switch format {
            case .json:
                let data = try CircleBoxSerializer.jsonData(from: envelope)
                urls.append(try fileStore.writeExportData(data, ext: "json"))
            case .csv:
                let data = CircleBoxSerializer.csvData(from: envelope)
                urls.append(try fileStore.writeExportData(data, ext: "csv"))
            case .jsonGzip:
                let data = try CircleBoxSerializer.jsonData(from: envelope)
                let compressed = try CircleBoxSerializer.gzipData(data)
                urls.append(try fileStore.writeExportData(compressed, ext: "json.gz"))
            case .csvGzip:
                let data = CircleBoxSerializer.csvData(from: envelope)
                let compressed = try CircleBoxSerializer.gzipData(data)
                urls.append(try fileStore.writeExportData(compressed, ext: "csv.gz"))
            case .summary:
                let data = try CircleBoxSerializer.summaryData(from: envelope, exportSource: envelope.exportSource.rawValue)
                urls.append(try fileStore.writeExportData(data, ext: "summary.json"))
            }
        }

        return urls
    }

    func hasPendingCrashReport() -> Bool {
        fileStore.hasPendingCrashReport()
    }

    func clearPendingCrashReport() throws {
        try fileStore.clearPendingCrashReport()
    }

    func debugSnapshot(maxEvents: Int = 200) -> [CircleBoxEvent] {
        let localConfig = stateQueue.sync { config }
        guard localConfig.enableDebugViewer else {
            return []
        }

        let events = ringBuffer.snapshot()
        let clamped = max(1, maxEvents)
        return Array(events.suffix(clamped))
    }

    func recoverPendingFromSignalMarkerIfNeeded() {
        if fileStore.hasPendingCrashReport() {
            try? fileStore.clearSignalMarker()
            return
        }

        guard let marker = fileStore.readSignalMarker() else {
            return
        }
        defer { try? fileStore.clearSignalMarker() }

        let baseEnvelope = fileStore.readCheckpointEnvelope() ?? snapshotEnvelope()
        let nextSeq = (baseEnvelope.events.last?.seq ?? -1) + 1
        let crashEvent = CircleBoxEvent(
            seq: nextSeq,
            timestampUnixMs: marker.timestampUnixMs,
            uptimeMs: Self.uptimeMs(),
            type: "native_exception_prehook",
            thread: .crash,
            severity: .fatal,
            attrs: [
                "signal": marker.name,
                "signal_number": String(marker.signal),
                "details": "signal:\(marker.name)(\(marker.signal))"
            ]
        )

        let recoveredEnvelope = CircleBoxEnvelope(
            schemaVersion: max(baseEnvelope.schemaVersion, 2),
            sessionId: baseEnvelope.sessionId,
            platform: baseEnvelope.platform,
            appVersion: baseEnvelope.appVersion,
            buildNumber: baseEnvelope.buildNumber,
            osVersion: baseEnvelope.osVersion,
            deviceModel: baseEnvelope.deviceModel,
            exportSource: .pendingCrash,
            captureReason: .startupPendingDetection,
            generatedAtUnixMs: Self.nowMs(),
            events: baseEnvelope.events + [crashEvent]
        )

        do {
            try fileStore.writePendingEnvelope(recoveredEnvelope)
            try? fileStore.clearCheckpointEnvelope()
        } catch {
            // Recovery must never interrupt SDK startup.
        }
    }

    private func handleCrash(details: String) {
        record(
            type: "native_exception_prehook",
            severity: .fatal,
            attrs: ["details": details],
            thread: .crash,
            persistCheckpoint: false
        )

        do {
            // Keep crash path best-effort and minimal: snapshot + single file write.
            let envelope = snapshotEnvelope(exportSource: .pendingCrash, captureReason: .uncaughtException)
            try fileStore.writePendingEnvelope(envelope)
            try? fileStore.clearSignalMarker()
        } catch {
            // Crash path should avoid throwing.
        }
    }

    private func snapshotEnvelope(
        exportSource: CircleBoxExportSource = .liveSnapshot,
        captureReason: CircleBoxCaptureReason = .manualExport
    ) -> CircleBoxEnvelope {
        let events = ringBuffer.snapshot()
        return CircleBoxEnvelope(
            sessionId: environment.sessionID,
            platform: environment.platform,
            appVersion: environment.appVersion,
            buildNumber: environment.buildNumber,
            osVersion: environment.osVersion,
            deviceModel: environment.deviceModel,
            exportSource: exportSource,
            captureReason: captureReason,
            generatedAtUnixMs: Self.nowMs(),
            events: events
        )
    }

    private func normalizedEnvelope(
        _ envelope: CircleBoxEnvelope,
        fallbackExportSource: CircleBoxExportSource,
        fallbackCaptureReason: CircleBoxCaptureReason
    ) -> CircleBoxEnvelope {
        let source = envelope.schemaVersion < 2 ? fallbackExportSource : envelope.exportSource
        let reason = envelope.schemaVersion < 2 ? fallbackCaptureReason : envelope.captureReason

        return CircleBoxEnvelope(
            schemaVersion: max(envelope.schemaVersion, 2),
            sessionId: envelope.sessionId,
            platform: envelope.platform,
            appVersion: envelope.appVersion,
            buildNumber: envelope.buildNumber,
            osVersion: envelope.osVersion,
            deviceModel: envelope.deviceModel,
            exportSource: source,
            captureReason: reason,
            generatedAtUnixMs: envelope.generatedAtUnixMs,
            events: envelope.events
        )
    }

    private func record(
        type: String,
        severity: CircleBoxEventSeverity,
        attrs: [String: String],
        thread: CircleBoxEventThread,
        persistCheckpoint: Bool = true
    ) {
        let activeConfig = stateQueue.sync { config }
        let sanitizedAttrs = CircleBoxSanitizer.sanitize(attrs: attrs, config: activeConfig)

        ringBuffer.append { seq in
            CircleBoxEvent(
                seq: seq,
                timestampUnixMs: Self.nowMs(),
                uptimeMs: Self.uptimeMs(),
                type: type,
                thread: thread,
                severity: severity,
                attrs: sanitizedAttrs
            )
        }

        if persistCheckpoint {
            writeCheckpointBestEffort()
        }
    }

    private func writeCheckpointBestEffort() {
        do {
            try fileStore.writeCheckpointEnvelope(snapshotEnvelope())
        } catch {
            // Checkpoint writes are best-effort and should not affect runtime behavior.
        }
    }

    private static func nowMs() -> Int64 {
        Int64(Date().timeIntervalSince1970 * 1000)
    }

    private static func uptimeMs() -> Int64 {
        Int64(ProcessInfo.processInfo.systemUptime * 1000)
    }
}

File: android/circlebox-sdk/src/main/kotlin/com/circlebox/sdk/CircleBoxRuntime.kt
package com.circlebox.sdk

import android.content.Context
import android.os.Looper
import android.os.SystemClock

internal class CircleBoxRuntime(
    context: Context
) {
    private val appContext = context.applicationContext
    private val lock = Any()

    private var config = CircleBoxConfig()
    private var environment = CircleBoxEnvironmentFactory.capture(appContext)
    private var ringBuffer = CircleBoxRingBuffer<CircleBoxEvent>(config.bufferCapacity)
    private var fileStore = CircleBoxFileStore(appContext)
    private var collectors: CircleBoxSignalCollectors? = null
    private var crashHandler: CircleBoxCrashHandler? = null
    private var started = false

    fun start(config: CircleBoxConfig) {
        synchronized(lock) {
            // Startup is idempotent: first successful call configures the runtime.
            if (started) {
                return
            }
            started = true
            this.config = config
            this.environment = CircleBoxEnvironmentFactory.capture(appContext)
            this.ringBuffer = CircleBoxRingBuffer(config.bufferCapacity)
            this.fileStore = CircleBoxFileStore(appContext)
        }

        val collectors = CircleBoxSignalCollectors(
            context = appContext,
            config = config,
            fileStore = fileStore,
            sink = { type, severity, attrs, thread ->
                record(type, severity, attrs, thread)
            }
        )
        this.collectors = collectors
        collectors.start()

        val crashHandler = CircleBoxCrashHandler { thread, throwable ->
            handleCrash(thread, throwable)
        }
        crashHandler.install()
        this.crashHandler = crashHandler

        record(
            type = "sdk_start",
            severity = CircleBoxEventSeverity.INFO,
            attrs = mapOf("buffer_capacity" to config.bufferCapacity.toString()),
            thread = CircleBoxEventThread.MAIN
        )
        writeCheckpointBestEffort()
    }

    fun breadcrumb(message: String, attrs: Map<String, String>) {
        val allAttrs = LinkedHashMap(attrs)
        allAttrs["message"] = message
        record(
            type = "breadcrumb",
            severity = CircleBoxEventSeverity.INFO,
            attrs = allAttrs,
            thread = currentThreadType()
        )
    }

    fun exportLogs(formats: Set<CircleBoxExportFormat>): List<java.io.File> {
        // Pending crash reports (from previous process runs) take precedence.
        val pendingEnvelope = fileStore.readPendingEnvelope()
        val envelope = normalizeEnvelope(
            pendingEnvelope ?: snapshotEnvelope(
                exportSource = CircleBoxExportSource.LIVE_SNAPSHOT,
                captureReason = CircleBoxCaptureReason.MANUAL_EXPORT
            ),
            fallbackExportSource = if (pendingEnvelope == null) CircleBoxExportSource.LIVE_SNAPSHOT else CircleBoxExportSource.PENDING_CRASH,
            fallbackCaptureReason = if (pendingEnvelope == null) CircleBoxCaptureReason.MANUAL_EXPORT else CircleBoxCaptureReason.STARTUP_PENDING_DETECTION
        )
        val exports = ArrayList<java.io.File>()

        if (formats.contains(CircleBoxExportFormat.JSON)) {
            exports += fileStore.writeExport(CircleBoxSerializer.encodeEnvelope(envelope), "json")
        }
        if (formats.contains(CircleBoxExportFormat.CSV)) {
            exports += fileStore.writeExport(CircleBoxSerializer.toCsv(envelope), "csv")
        }
        if (formats.contains(CircleBoxExportFormat.JSON_GZIP)) {
            val raw = CircleBoxSerializer.encodeEnvelope(envelope)
            exports += fileStore.writeExport(CircleBoxSerializer.gzip(raw), "json.gz")
        }
        if (formats.contains(CircleBoxExportFormat.CSV_GZIP)) {
            val raw = CircleBoxSerializer.toCsv(envelope)
            exports += fileStore.writeExport(CircleBoxSerializer.gzip(raw), "csv.gz")
        }
        if (formats.contains(CircleBoxExportFormat.SUMMARY)) {
            val summary = CircleBoxSerializer.encodeSummary(envelope, exportSource = envelope.exportSource.name.lowercase())
            exports += fileStore.writeExport(summary, "summary.json")
        }

        return exports
    }

    fun hasPendingCrashReport(): Boolean = fileStore.hasPendingCrashReport()

    fun clearPendingCrashReport() {
        fileStore.clearPendingCrashReport()
    }

    fun debugSnapshot(maxEvents: Int): List<CircleBoxEvent> {
        val localConfig = synchronized(lock) { config }
        if (!localConfig.enableDebugViewer) {
            return emptyList()
        }

        val events = ringBuffer.snapshot()
        val clamped = maxOf(1, maxEvents)
        return events.takeLast(clamped)
    }

    private fun handleCrash(thread: Thread, throwable: Throwable) {
        record(
            type = "native_exception_prehook",
            severity = CircleBoxEventSeverity.FATAL,
            attrs = mapOf(
                "thread" to thread.name,
                "throwable" to throwable.javaClass.name,
                "message" to (throwable.message ?: "unknown")
            ),
            thread = CircleBoxEventThread.CRASH,
            persistCheckpoint = false
        )

        runCatching {
            // Keep crash path best-effort and minimal.
            fileStore.writePendingEnvelope(
                snapshotEnvelope(
                    exportSource = CircleBoxExportSource.PENDING_CRASH,
                    captureReason = CircleBoxCaptureReason.UNCAUGHT_EXCEPTION
                )
            )
        }
    }

    private fun snapshotEnvelope(
        exportSource: CircleBoxExportSource = CircleBoxExportSource.LIVE_SNAPSHOT,
        captureReason: CircleBoxCaptureReason = CircleBoxCaptureReason.MANUAL_EXPORT
    ): CircleBoxEnvelope {
        val env = environment
        return CircleBoxEnvelope(
            sessionId = env.sessionId,
            platform = env.platform,
            appVersion = env.appVersion,
            buildNumber = env.buildNumber,
            osVersion = env.osVersion,
            deviceModel = env.deviceModel,
            exportSource = exportSource,
            captureReason = captureReason,
            generatedAtUnixMs = nowMs(),
            events = ringBuffer.snapshot()
        )
    }

    private fun normalizeEnvelope(
        envelope: CircleBoxEnvelope,
        fallbackExportSource: CircleBoxExportSource,
        fallbackCaptureReason: CircleBoxCaptureReason
    ): CircleBoxEnvelope {
        val source = if (envelope.schemaVersion < 2) fallbackExportSource else envelope.exportSource
        val reason = if (envelope.schemaVersion < 2) fallbackCaptureReason else envelope.captureReason

        return envelope.copy(
            schemaVersion = maxOf(2, envelope.schemaVersion),
            exportSource = source,
            captureReason = reason
        )
    }

    private fun record(
        type: String,
        severity: CircleBoxEventSeverity,
        attrs: Map<String, String>,
        thread: CircleBoxEventThread,
        persistCheckpoint: Boolean = true
    ) {
        val localConfig = synchronized(lock) { config }
        val sanitized = CircleBoxSanitizer.sanitize(attrs, localConfig)

        ringBuffer.append { seq ->
            CircleBoxEvent(
                seq = seq,
                timestampUnixMs = nowMs(),
                uptimeMs = SystemClock.elapsedRealtime(),
                type = type,
                thread = thread,
                severity = severity,
                attrs = sanitized
            )
        }

        if (persistCheckpoint) {
            writeCheckpointBestEffort()
        }
    }

    private fun writeCheckpointBestEffort() {
        runCatching { fileStore.writeCheckpointEnvelope(snapshotEnvelope()) }
    }

    private fun currentThreadType(): CircleBoxEventThread {
        return if (Looper.myLooper() == Looper.getMainLooper()) {
            CircleBoxEventThread.MAIN
        } else {
            CircleBoxEventThread.BACKGROUND
        }
    }

    private fun nowMs(): Long = System.currentTimeMillis()
}
